### PR TITLE
ddl: integrate fast create table into normal general DDL workflow

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #54436

Problem Summary:

### What changed and how does it work?
- as title, we reuse the variable to enable merging create table jobs in normal general DDL. reason:

Compared with current general DDL, the speedup of fast-create-table mostly due to merging multiple CreateTable jobs into 1 CreateTables job, and running them together and disable wait schema version synchonization.  So we have to use a larger concurrency to make them merge as we are merging from the job channel. Without merging jobs, such as without enough concurrency, it's even slower than current general DDL, as it has less workers in most cases. **Normal general DDL can merge jobs too**.
current fast-create-table is harder to support creating tables with foreign keys without wait schema version change, and without waiting for schema version synced, it causes schema inconsistency for half ddl lease which is expected. **Normal general DDL don't have such issues**.

- some code of previous fast-create is still there, will remove them later.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

same case in https://github.com/pingcap/tidb/pull/54693, with fast-create flag enabled, takes 9m4. when run with 100 threads, takes 4m20
image

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/executor.go [hunk 1]
```diff
@@ -10105,20 +10111,29 @@ func (e *executor) notifyNewJobByEtcd(etcdPath string, jobID int64, jobType stri
...
-		e.limitJobChV2 <- task
-	} else {
-		e.limitJobCh <- task
-	}
+	// TODO this might block forever, as the consumer part considers context cancel.
+	e.limitJobCh <- task
 }
 
 // addBatchDDLJobsV1 gets global job IDs and puts the DDL jobs in the DDL queue.
 func (d *ddl) addBatchDDLJobsV1(jobWs []*JobWrapper) {
```
Comment (GMHDBJD) [inline pkg/ddl/executor.go:10119]: maybe we can revert the name to `addBatchDDLJobs` in later pr.

File: pkg/ddl/executor.go [hunk 2]
```diff
@@ -10139,19 +10154,20 @@ func (d *ddl) addBatchDDLJobsV1(jobWs []*JobWrapper) {
...
 			zap.Int("batch count", len(jobWs)),
 			zap.String("jobs", jobs),
-			zap.Bool("table", toTable))
+			zap.Bool("table", toTable),
+			zap.Bool("fast_create", fastCreate))
 	}
 }
 
 // addBatchLocalDDLJobs gets global job IDs and delivery the DDL jobs to local TiDB
 func (d *ddl) addBatchLocalDDLJobs(jobWs []*JobWrapper) {
```
Comment (GMHDBJD) [inline pkg/ddl/executor.go:10163]: will not be called currently?

**Review by GMHDBJD — COMMENTED** [review]
rest LGTM

File: pkg/ddl/executor.go (hunk 1 above)
Comment (D3Hunter) [reply pkg/ddl/executor.go:10119]: lgtm

File: pkg/ddl/executor.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/ddl/executor.go:10163]: yes, some code are not removed currently, as this pr is quite large, will do it later

**Review by GMHDBJD — APPROVED** [review]
LGTM, remember to update docs

File: pkg/session/session.go [hunk 3]
```diff
@@ -3221,12 +3221,7 @@ func InitDDLJobTables(store kv.Storage, targetVer meta.DDLTableVersion) error {
 		targetTables = BackfillTables
 	}
 	return kv.RunInNewTxn(kv.WithInternalSourceType(context.Background(), kv.InternalTxnDDL), store, true, func(_ context.Context, txn kv.Transaction) error {
-		var t *meta.Meta
-		if variable.EnableFastCreateTable.Load() {
-			t = meta.NewMeta(txn, meta.WithUpdateTableName())
-		} else {
```
Comment (tangenta) [inline pkg/session/session.go:3227]: Is there a plan to clean these `tableName -> tableID` kvs on old cluster?

Comment (D3Hunter) [reply pkg/session/session.go:3227]: fast create is expeironmental, we will write in doc to let customer disable it before upgrade

Comment (D3Hunter) [reply pkg/session/session.go:3227]: https://github.com/pingcap/docs-cn/pull/18214 doc pr

File: pkg/ddl/executor.go [hunk 4]
```diff
@@ -10524,46 +10511,64 @@ func lockGlobalIDKey(ctx context.Context, ddlSe *sess.Session, txn kv.Transactio
...
-		} else if jobW.Job.SchemaName != schemaName {
-			return jobWs, nil
+		// ActionCreateTables doesn't support foreign key now.
+		tbInfo, ok := jobW.Args[0].(*model.TableInfo)
+		if !ok || len(tbInfo.ForeignKeys) > 0 {
+			resJobWs = append(resJobWs, jobW)
+			continue
 		}
-		jobs = append(jobs, jobW.Job)
+		// CreateTables only support tables of same schema now.
```
Comment (tangenta) [inline pkg/ddl/executor.go:10533]: Why do we have this limitation?

Comment (D3Hunter) [reply pkg/ddl/executor.go:10533]: initial design has this limit https://github.com/pingcap/tidb/pull/28763

we want to extend it to support tables of different schema, but requires CDC to support it too, maybe in 8.4 or later

**Review by tangenta — APPROVED** [review]

**Review by yudongusa — APPROVED** [review]
LGTM

**Pull request merged**